components/constants/brave_services_key_v2.cc

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/constants/brave_services_key_v2.h"
+
+#include <vector>
+#include "base/logging.h"
+
+#include "base/base64.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/stringprintf.h"
+#include "brave/components/constants/buildflags.h"
+#include "crypto/hmac.h"
+#include "crypto/sha2.h"
+
+namespace constants {
+
+namespace {
+
+std::optional<std::string> GetServiceKey(Service service) {
+  if (service == Service::kAIChat) {
+    return BUILDFLAG(AI_CHAT_SERVICE_KEY);
+  }
+
+  return std::nullopt;
+}
+
+}  // namespace
+
+std::optional<std::pair<std::string, std::string>> GetBraveServicesV2Headers(
```

Comment:
if we want to make this re-usable for other services we need to account for the fact that other services might pass a different list of headers ( and pass the full request to this function. ) it's also probably a good idea to break out the signature string generation separately into it's own function.

see https://github.com/brave-intl/bat-go/blob/master/libs/httpsignature/httpsignature.go#L111-L167 for a full example implementation, but at the least we should consider using a similar function signature.

Reply:
https://github.com/brave/brave-core/blob/ad741ad411c21681993a7266546eb394bbb5d1a8/components/brave_rewards/core/common/request_signer.h seems to contain existing browser code we could adapt ( maybe extract out to top level crypto utils? )

Reply:
Sounds like it might be better to call the signing code from within APIRequestHelper::Request, having passed some APIRequestOptions that instructs which headers to sign / which key_id and services key to use @yrliou?

I could create a RequestSigner implementation modeled after that one from rewards. Unifying them might be a little tricky given our the timeline

Reply:
Here's how I would approach it:
Have a components/brave_service_keys component or some other naming which makes sense (constants as the component name is a bit weird in my opinion) which will have functions for generating keys and signing following the spec and unit tests for each function. Better to have smaller functions rather than just one like here.
Specific service logic, such as how AI chat header would look like, should not be in this component.
How header is structured should live in its own component (in ai_chat for this particular case), if there are something worth sharing across components, create generalized util functions and put inside brave_service_keys component.